Add C++ function to accelerate DTensor.__new__

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #161695
* #161694
* #161693
* #161692
* #161635
* #161634
* #161633
* #161622
* #161596
* #161595
* #161591
* #161590
* __->__ #161588
* #161586
* #161466
* #161455
* #161432
* #161329
* #161328
* #161317
* #161315
* #161308
* #161304
* #161292
* #161301
* #161300
* #161286

As the new comment explains, pybinding DispatchKeySet is expensive; one native function avoids the number of DispatchKeySets we incur pybind overhead for to 1.

Reviewers please advise on a better location for the new torch._C function; I just threw it in Module.cpp because I didn't know where to put it.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta

swolchok committed:
Add C++ function to accelerate DTensor.__new__

As the new comment explains, pybinding DispatchKeySet is expensive; one native function avoids the number of DispatchKeySets we incur pybind overhead for to 1.

Reviewers please advise on a better location for the new torch._C function; I just threw it in Module.cpp because I didn't know where to put it.

[ghstack-poisoned]

swolchok commented:
It might be better to write a specialized torch.Tensor._make_dtensor variant of _make_wrapper_subclass. Will look into that next. Can do this as an incremental step though.

swolchok commented:
> It might be better to write a specialized torch.Tensor._make_dtensor variant of _make_wrapper_subclass. Will look into that next. Can do this as an incremental step though.

yeah, that approach speeds up benchmark significantly, will put it up soonish

ezyang reviewed (approved):
By the way, most wrapper subclasses need to handle negative/conjugate correctly (DTensor had a fix bodged on just for it because I was lazy), so there might be a bigger change that eliminates the problem for all subclass users without another one off function

swolchok committed:
Update on "Add C++ function to accelerate DTensor.__new__"


As the new comment explains, pybinding DispatchKeySet is expensive; one native function avoids the number of DispatchKeySets we incur pybind overhead for to 1.

Reviewers please advise on a better location for the new torch._C function; I just threw it in Module.cpp because I didn't know where to put it.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]
